### ISSUE TITLE
Update Maven packages

### ISSUE THREAD

**Issue by DanielOleo** [javagl/JglTF#59]
Hello,

I'm looking to include this library in a project I'm working currently on, however the packages currently on Maven are reporting a bunch of security dependancy vulnerabilities. Plus the package seems to be behind the current code base

Is this something that could easily be updated?

https://mvnrepository.com/artifact/de.javagl/jgltf-model/2.0.0

**Comment by javagl** [comment]
I could basically update it "at any point in time" (I would need to allocate some time for that, but that should be doable). An update has also been requested in https://github.com/javagl/JglTF/issues/54 , and in the meantime, there actually _have_ been some changes and additions that could justify a new release. This includes smaller fixes, as well as the (somewhat experimental) https://github.com/javagl/JglTF/tree/master/jgltf-model-builder project. The latter is something that I'm still actively working on. (So strictly speaking, it should be in a branch, but I moved an early "proof of concept" state of it into `master`). 

I'm a bit on the fence with a detail: The version number. Within the concept of semantic versioning, this would have to be `3.0.0`, because there certainly have been some breaking changes. But there should be _few_ changes that will _actually_ require changes on the client side (except for clients that used classes that have clearly been marked as "Not part of the public API" via the JavaDocs). It might also be called `2.1.0`, with some disclaimers...

In any case: The security vulnurabilities are, an many ways, a **strong** reason to do an update ASAP - **at least** including the updated dependency versions (and that could be a `2.0.1`, actually). But it could make sense to combine that with a "reasonable last state" of the code base, including the other changes.

If anybody has any preferences, just chime in here. (Maybe with a ping @jailln who opened the other issue). 

Otherwise, I'll try to work towards a release ASAP, and think about details like the version number, or whether it should already contain the `jgltf-model-builder` (with the disclaimer that it is still experimental), and the overall timeline for that.

**Comment by jailln** [comment]
No strong opinions on this but if this is not too much work I would suggest doing a release with the security vulnerabilities corrections first ( a `2.0.1`)  and then release a `2.1.0` with the other modifications (maybe with the `jgltf-model-builder` with a disclaimer depending on its state and on the benefits of having it) and finally a `3.0.0` when `jgltf-model-builder` is fully operational. But again I don't know the project well enough to have a clear opinion :)

**Comment by DanielOleo** [comment]
No strong opinion here on version numbers either, what been suggested seems good

I hadn't seen the model builder, l've been working through model creation the long way! So that would be a welcome addition when its complete

I would found this piece of code from the tinygltf project quite a handy guide on building a GLTF file from scratch https://github.com/syoyo/tinygltf/blob/master/examples/build-gltf/create_triangle_gltf.cpp

Might be a demo worth replicating at some point, even as just a test case

**Comment by javagl** [comment]
@jailln 

> No strong opinions on this but if this is not too much work I would suggest doing a release with the security vulnerabilities corrections first ( a `2.0.1`)  and then release a `2.1.0` 

That definitely makes sense - just to have the compatible version, without the vulnerabilities in the depdendencies. I just did a release 2.0.1 with the latest dependency versions for Jackson and JUnit. Actually, I hadn't done a Maven release from a tag/branch before, but it _seemed_ to work. It's already in https://repo1.maven.org/maven2/de/javagl/jgltf-model/2.0.1/ and will probably show up on the search site when some indices are updated (and this time, hopefully without vulnerabilities). (If it does not show up tomorrow, something might be wrong, then I'll have another look....). 

Some projects from this state have been split, to it wouldn't make sense to cherry-pick/merge this change back to `master`, but I'll just did this manually...

---

@DanielOleo 

>  l've been working through model creation the long way! So that would be a welcome addition when its complete

Well, "complete" is one of these difficult words ... ;-) 

But if you went through this, and came up with ideas ("use-cases"), along the lines of "I'd like to create **X**, and this is the `code` that I'd like to write for that", then any feedback that might find its way into the "model builder" would be appreciated.

(It's hard to give a definite timeline, though. All this is just some spare-time stuff, and I sometimes wonder why I'm doing that at all...) 

The `create_triangle_gltf.cpp` shows that _very_ low-level approach of building a model. The code has ~120 lines, and the glTF asset that it creates has ~70 lines. And I could ask: Who really wants to manually define floating-point values as blocks of `0x3f` bytes, and manually include padding, and manually set indices? **But** this is not meant to be dismissive: It certainly makes sense to _have_ this low-level layer. People can build other cool stuff on top of that. 

(This "low-level" layer roughly corresonds to the `jgltf-impl-v2` project: You _could_, if you wanted to, go the whole (painful) way of manually creating `Accessor` and `BufferView` objects, assigning indices, fight with low-level spec details like byte stride and so on...)

But ... I don't want to manually set indices. I don't want to manually encode buffers. I don't want to manually insert padding bytes. I don't want to manually set "default" values. Therefore, I wanted to create this "model builder" on the level of the `gltf-model` classes. (They are not perfect. I'm definitely not satisfied with how they played out. But ... let's ignore that for now).

As an example, I just added a (small) test for the current state of the `model-builder` project: The [`GltfModelCreationBasicExample.java`](https://github.com/javagl/JglTF/blob/0a6897777e84f6bf7183146b30f369d00e0137d4/jgltf-model-builder/src/test/java/de/javagl/jgltf/model/creation/example/GltfModelCreationBasicExample.java#L31) shows the code that essentially corresponds to the `create_triangle_gltf.cpp`, in that it creates an embedded glTF asset with a single triangle.

The possibility to simply say
`meshPrimitiveBuilder.setIntIndicesAsShort(IntBuffer.wrap(indices));`
to just set the indices from an `int[]` array, or something like 
`materialBuilder.setBaseColorTexture("./data/testTexture.png", "testTexture.png", null);`
(from the other example) to add some texture, without having to think about indices or byte offsets, should make it easier to create some basic models. 

**But...** ... of course, as usual, this sort of convenience comes at the (potential) price of lower flexibility. I'm trying to head towards the goal: "Simple things should be _simple_, and complex things should be _possible_". Let's see in how far I can achieve that....

**Comment by DanielOleo** [comment]
The interface demo'd in GltfModelCreationBasicExample.java would definitely be a welcome one

> (This "low-level" layer roughly corresonds to the jgltf-impl-v2 project: You could, if you wanted to, go the whole (painful) way of manually creating Accessor and BufferView objects, assigning indices, fight with low-level spec details like byte stride and so on...)

^ This is exactly what I am doing, it's painful, but do-able. The meshes I'm creating (for the moment) are simple extrusions of shapes previously written in OBJ/MTL format instead, so I'm not using the full range of what GLTF has to offer **_yet_**, so I like the idea of "Simple things should be simple, and complex things should be possible"!

**Comment by javagl** [comment]
Something that had previously been written as OBJ should already be covered by the current functionality. 

(In fact, I changed the `jgltf-obj` project at some point to use the new structures (although the [corresponding commit](https://github.com/javagl/JglTF/commit/ae827b0113a950b3378f653796fdd2d2688c255f) will probably not be so useful for you without context...)

I assume that the results are just "single scene, single node, single mesh, single mesh primitive" assets, where the latter contains the (former OBJ) geometry data. There still _are_ some degrees of freedom even for that. This mainly refers to the overall data layout: Should there be one `bufferView` for each `accessor`, or one `buffer` for each `bufferView`...? I tried to "prepare" some configuarbility here via the `BufferBuilderStrategy` interface with its [`DefaultBufferBuilderStrategy`](https://github.com/javagl/JglTF/blob/master/jgltf-model-builder/src/main/java/de/javagl/jgltf/model/creation/DefaultBufferBuilderStrategy.java). But this is not yet completely thought through and settled. 

And it might not be important in many cases. At least, I could imagine that people don't care about the layout, as long as the result is a "somewhat reasonable (and _valid_)" glTF asset. But I'll probably consider some further options here, and see how they could be exposed on the API level.

**Comment by javagl** [comment]
I'm starting to worry why https://mvnrepository.com/artifact/de.javagl/jgltf-model still does not show 2.0.1. But it's already in https://repo1.maven.org/maven2/de/javagl/jgltf-model/2.0.1/ , so I'll give it another day, who knows how often the indices are updated...

**Comment by DanielOleo** [comment]
Yep, I see it on my side. Thanks for sorting :)

I'll take a look at the changes in `jgltf-obj`. As we move into more complex file creation the more the library can do for us the better obviously. So I look forward to using these additions one day (when/if you find enough spare-time to get it to a API stable place etc.)

Thanks for the advice,

**Comment by javagl** [comment]
Again, the _changes_ in the _specific commit _are probably not so helpful. Roughly speaking: The `jgltf-obj` contained _some_ classes that could take the geometry data (vertices, indices, normals...) from the OBJ, and create a glTF asset that contained _exactly this_ geometry. I generalized this in the `jgltf-model-builder`, and the commit that I linked to basically replaced the OBJ-specific classes for building glTF assets with the generalized ones from `jgltf-model-builder`. 

But the question of the API is not fully settled. As one, specific example: There are convenience functions like
`meshPrimitiveBuilder.setIntIndicesAsShort(IntBuffer.wrap(indices));`
for the common case that someone has an `int[]` array with indices, and wants to use them as (`GL_UNSINED_SHOT`) indices in 

the glTF asset. But there should _still_ be the possibility to assign "arbitrary data" to the mesh primitive - for example, some 3x3-matrices with normalized `short` values:
```
AccessorModel normalizedShort3x3Matrices = 
    AccessorModels.create(GL_SHORT, "MAT3", true, rawData);
meshPrimitiveBuilder.addAttribute(
    "_MY_ATTRIBUTE", normalizedShort3x3Matrices);
```
All this is in the process of being sorted out. I have some "use-cases", and try to derive possible requirements from that - but it's still happening without much feedback from the outside world...

**Comment by javagl** [comment]
I'd be tempted to close this, because the issue for `2.0.0` was resolved by publishing `2.0.1`, ... but ... a  vulnurability has been discovered for a dependency of this state was well (why is it always `jackson-databind`... ). I'll try to schedule an update, and ... leave this issue open for now...

**Comment by javagl** [comment]
I recently updated the dependencies, and just released `2.0.2`, so I'll close this one for now (until the next `jackson-databind` issue is discovered...)
